### PR TITLE
Python cli

### PR DESCRIPTION
**Pull request by oliver-sanders**
> **Important:**
>
> Before this PR each command was installed individually as it's own script.
>
> `pip` is good at installing scripts, but not quite so good at uninstalling them.
> For maximum safety either use a clean environment OR manually remove all rose scripts
> (scripts no longer listed in setup.py may accumulate in your envs bin dir otherwise).
>
> Activate your env then use `which rose` to locate your rose scripts then do `rm /path/to/rose/*`.
> Repeat for `rosie` and `rosa`.
>
> After this change there are only three scripts, `rose`, `rosie` and `rosa`. All sub commands
> are implemented as entry points or in the case of bash commands manually registered with
> the root command.

Trying to keep this changeset to the minimal diff needed to switch the CLI to Python. The CLI isn't built in the nicest way and required some optparse formatting shims to preserve formatting. We can improve this later if desired.

* cli: remove shell wrapper scripts
    * Python sub-commands are now Python entry points executed in the same
      Python session (and not via an intermediate shell).
    * Bash sub-commands are now located in metomi.rose.scripts (incl rosa
      sub-commands) and called via subprocess.Popen.
    * Re-work the Sphinx extension used to auto-document the CLI to work
      with the new CLI format.
* check-software: migrate into metomi library
    * Move this lone Python script from bin/ into metomi/rose and register
      it as an entry point.
* rosa: migrate rosa scripts into bin dir
    * Rosa scripts used to be located in sbin/ they are now merged in with
      the remaining non-Python scripts.
    * The top-level Rosa command is only installed if the Python optional
      dependency is specified.
* docs: minor changes to documentation
    * Fix RST.
    * Remove references to rosie-go.
    * Bump cylc-flow to 8.0b2 via intersphinx.
    * Add deprecation warnings to obsolete (but still documented) commands.
    * Fix Sphinx errors.
* cli: migrate from Bash to Python
    * Convert the main Rose script from Bash to Python.
    * Register Python commands as entry points, hardcode Bash commands
      in the main Rose script.
    * Update centralised option help text to the more up to date
      help text in the sub commands.

### CODE REVIEW

File: metomi/rose/opt_parse.py [hunk 1]
```diff
@@ -29,22 +86,32 @@ class RoseOptionParser(OptionParser):
...
             },
         ],
         "all_revs": [
             ["--all-revs"],
             {
                 "action": "store_true",
                 "default": False,
-                "help": "Return all revisions of matched items.",
+                "help": (
+                    "Specify whether to search deleted suites and superceded"
```
Comment (wxtim) [inline metomi/rose/opt_parse.py:112]:
```suggestion
                    "Specify whether to search deleted suites and superseded"
```
https://www.collinsdictionary.com/dictionary/english/supercede

File: metomi/rose/opt_parse.py [hunk 2]
```diff
@@ -215,7 +327,10 @@ class RoseOptionParser(OptionParser):
                 "action": "store",
                 "dest": "diff_tool",
                 "default": None,
-                "help": "Specify an alternate diff tool like diffuse.",
+                "help": (
+                    "Specify an alternate diff tool."
+                    "\nE.G: diffuse, vimdiff or kompare."
```
Comment (wxtim) [inline metomi/rose/opt_parse.py:332]: If you wrote "Exempli gratia" in full you'd have

```suggestion
                    "\nE.g: diffuse, vimdiff or kompare."
```
or you could just skip the Latin?
```suggestion
                    "\nFor example: diffuse, vimdiff or kompare."
```

File: setup.cfg [hunk 3]
```diff
@@ -41,3 +41,40 @@ classifiers =
...
+    macro = metomi.rose.macro:main
+    metadata-check = metomi.rose.metadata_check:main
+    metadata-gen = metomi.rose.metadata_gen:main
+    metadata-graph = metomi.rose.metadata_graph:main
+    namelist-dump = metomi.rose.namelist_dump:main
+    resource = metomi.rose.resource:main
+    task-env = metomi.rose.task_env:main
+    task-run = metomi.rose.task_run:main
+rosie.commands =
+    checkout = metomi.rosie.vc:checkout
```
Comment (wxtim) [inline setup.cfg:71]: What's happened to `rosie-co` - I know it's only an alias, but...

File: sphinx/api/command-reference.rst [hunk 4]
```diff
@@ -16,21 +16,38 @@ Rose Commands
...
+   The old Rose 2019 (Python 2) GUI remains compatible with Rose 2
+   configurations.
 
 .. _command-rose-suite-run:
 
 rose suite-run
 ^^^^^^^^^^^^^^
 
-TODO: This is here to allow the documentation tests to pass
+This command has been replaced by ``cylc install ; cylc play``.
```
Comment (wxtim) [inline sphinx/api/command-reference.rst:31]: Refraining from making further comment on these sections given the `.. TODO: ` below. Happy to make suggestions now, or afterwards if you wish.

File: sphinx/conf.py [hunk 5]
```diff
@@ -77,7 +77,7 @@
 
 # mapping to other Sphinx projects
 # (allows us to reference objects from other projects)
-cylc_version = '8.0a2'
+cylc_version = '8.0b2'
```
Comment (wxtim) [inline sphinx/conf.py:80]: Will this be OK if testing with 8.0rc1?

**Review by wxtim — COMMENTED** [review]
Had a read through, checking the old vs. the new interfaces. Will check out and have a play once tests are passing.

File: setup.cfg (hunk 3 above)
Comment (oliver-sanders) [reply setup.cfg:71]: Aliases are still there, they are hardcoded in `metomi.rose.rose`.

File: sphinx/api/command-reference.rst (hunk 4 above)
Comment (oliver-sanders) [reply sphinx/api/command-reference.rst:31]: I started making docs changes but backed out to do one thing at a time.

We need to remove `rose suite-run` references from the docs next :grimacing:.

File: sphinx/conf.py (hunk 5 above)
Comment (oliver-sanders) [reply sphinx/conf.py:80]: This is version used by intersphinx which is templated into the `cylc-doc` url for retrieving Sphinx objects.

This means we can do things like `:term:"cycle point"` to reference a term in the Cylc glossary.

The version used must be published so we will need to pin to the last release.

File: metomi/rose/opt_parse.py (hunk 2 above)
Comment (oliver-sanders) [reply metomi/rose/opt_parse.py:332]: Capitalisation of e.g is mixed in all repos, not planning on addressing that right now.

**Comment by oliver-sanders** [comment]
Rebased, deconflicted and fixed tests, should be good to go now 🤞.

**Comment by oliver-sanders** [comment]
Nope.

**Comment by oliver-sanders** [comment]
One failing test for me to work on, otherwise fine.

File: metomi/rose/date_cli.py [hunk 6]
```diff
@@ -187,6 +185,10 @@ def main():
         file=sys.stderr
     )
 
+    if '--help' in sys.argv:
```
Comment (wxtim) [inline metomi/rose/date_cli.py:188]: `rose config-diff` on it's own returns `Usage: rose config-diff [options]` - fine as far as it goes, but shouldn't it return either the same as `--help` or a note to the effect of "see --help for details"?

File: metomi/rose/config_diff.py [hunk 7]
```diff
@@ -135,7 +135,73 @@ def format_metadata_as_text(metadata, only_these_options=None):
...
+
+       in the same location. This will allow you to run::
+
+          rose config-diff --ignore=foo ...
+
+       instead of::
+
+          rose config-diff --ignore=namelist:bar --ignore=namelist:baz ...
+        '''
+    )
```
Comment (wxtim) [inline metomi/rose/config_diff.py:204]: `rose config-diff` on it's own returns `Usage: rose config-diff [options]` - fine as far as it goes, but shouldn't it return either the same as `--help` or a note to the effect of "see --help for details"?

Comment (oliver-sanders) [reply metomi/rose/config_diff.py:208]: Good spot, was missing a usage line, fixed.

**Comment by wxtim** [comment]
Do we want to leave `rose date` (with it's deprecation warning) as is, or do we want a harder deprecation?

File: metomi/rose/env_cat.py [hunk 8]
```diff
@@ -25,8 +25,34 @@
 
 def main():
     """Implement "rose env-cat"."""
-    opt_parser = RoseOptionParser()
+    opt_parser = RoseOptionParser(
```
Comment (wxtim) [inline metomi/rose/env_cat.py:28]: 1. For another day - It took me a long time to understand this docstring (and I'm not sure I have).
2. What behavior should I see if I just execute the command with no opts or args? Is just hanging desirable?

File: metomi/rose/host_select_client.py [hunk 9]
```diff
@@ -22,7 +22,40 @@
 import psutil
 
 
+def usage():
```
Comment (wxtim) [inline metomi/rose/host_select_client.py:25]: This one also appears to hang with no args or opts.

File: metomi/rose/host_select_client.py [hunk 10]
```diff
@@ -22,7 +22,40 @@
...
+
+Internal command for obtaining information about hosts using the
+Python psutil module.
+
+The names of psutil methods are passed to stdin along with any arguments
+as a JSON list.
+
+EXAMPLES:
+    # return the virtual memory and cpu usage
+    [["virual_memory"], ["cpu_percent"]]
```
Comment (wxtim) [inline metomi/rose/host_select_client.py:37]:
```suggestion
    [["virtual_memory"], ["cpu_percent"]]
```

File: metomi/rose/host_select_client.py [hunk 11]
```diff
@@ -22,7 +22,40 @@
...
+
+    # return disk usage for the filesystem mounted at "/"
+    [["disk_usage", "/"]]
+
+The input json should be preceded by \n**start**\n
+and followed by \n**end**\n to avoid issues with user profile scripts
+and stdin deadlock.
+
+$ rose host-select-client <<'__HERE__'
```
Comment (wxtim) [inline metomi/rose/host_select_client.py:46]: I dont think that this worked when I tried it...

File: metomi/rose/metadata_gen.py [hunk 12]
```diff
@@ -141,8 +141,31 @@ def type_gen(value):
...
 
 def main():
-    opt_parser = metomi.rose.opt_parse.RoseOptionParser()
+    opt_parser = metomi.rose.opt_parse.RoseOptionParser(
+        usage='rose metadata-gen [OPTIONS] [PROPERTY=VALUE ...]',
+        description='''
+Automatically generate metadata from an application or suite
+configuration. An aid for metadata development.
+
+WARNING: May Contain Thorns.
```
Comment (wxtim) [inline metomi/rose/metadata_gen.py:150]: _really ?_

File: metomi/rose/rose.py [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+
+# fmt: off
+DEAD_ENDS = {
+    # messages to show for commands which have been removed or renamed
+    # (ns, sub_cmd): message
+    ('rose', 'config-edit'):
+        'The Rose configuration editor has been removed, use the Cylc GUI.',
+    ('rose', 'edit'):
+        'The Rose configuration editor has been removed, use the Cylc GUI.',
+    ('rose', 'metadata-graph'):
```
Comment (wxtim) [inline metomi/rose/rose.py:49]: Should this still have a script if it's in this dict?

File: metomi/rose/scripts/rose-tutorial [hunk 14]
```diff
@@ -28,9 +28,16 @@
 #
 # DESCRIPTION
 #     Make a copy of one of the tutorial SUITE in the cylc-run directory.
-#
 #-------------------------------------------------------------------------------
 set -eu
+
+if [[ $1 == '--help' ]]; then
```
Comment (wxtim) [inline metomi/rose/scripts/rose-tutorial:34]:
```suggestion
if [[ "${1:-}" == '--help' ]]; then
```

File: metomi/rose/scripts/rose-tutorial [hunk 15]
```diff
@@ -28,9 +28,16 @@
...
+    # shellcheck source=metomi/rose/etc/lib/bash/rose_usage
+    . "$(rose resource lib/bash/rose_usage)"
+    rose_help
+    exit 0
+fi
+
 mkdir -p "$HOME/cylc-run"
 
 usage () {
```
Comment (wxtim) [inline metomi/rose/scripts/rose-tutorial:43]: This doesn't work for me - at Rose 1 line 47 was   
```bash
ls -1 "${ROSE_HOME}/etc/tutorial/" | sed -e 's/^/    /'
```

I'm not sure how this find command is supposed to produce the desired result. Going digging into the change history.

File: metomi/rose/host_select_client.py (hunk 9 above)
Comment (oliver-sanders) [reply metomi/rose/host_select_client.py:25]: This is an internal command which waits for input from stdin.

The best we can do is abort if stdin is not received within a hardcoded period.

File: metomi/rose/host_select_client.py (hunk 11 above)
Comment (oliver-sanders) [reply metomi/rose/host_select_client.py:46]: Needed to remove the `$PS2` char when copying into the CLI.

File: metomi/rose/metadata_gen.py (hunk 12 above)
Comment (oliver-sanders) [reply metomi/rose/metadata_gen.py:150]: This command attempts to guess the type of inputs. It doesn't always give the right answer.

**Comment by wxtim** [comment]
rosie and rosa both have a slight top-level doc issue - under the list of commands they are all `rose rosie-command`.
```
> rosa
usage: rosa [--help] [--version]

Rosa 2.0b3: Admin utilities for Rosie.

optional arguments:
  --help, -h
  --version

Commands:
  rose db-create
  rose rpmbuild
  rose svn-post-commit
  rose svn-pre-commit
  rose ws
```

File: metomi/rose/scripts/rose-tutorial (hunk 14 above)
Comment (oliver-sanders) [reply metomi/rose/scripts/rose-tutorial]: Fixed in latest push.

File: metomi/rose/rose.py (hunk 13 above)
Comment (oliver-sanders) [reply metomi/rose/rose.py:49]: Awaiting reimplementation, not sure we have an issue for this.

Comment (oliver-sanders) [reply metomi/rose/rose.py:49]: https://github.com/metomi/rose/issues/2501

**Comment by wxtim** [comment]
Rosa rpmbuild is broken - have created issue #2502

**Review by wxtim — APPROVED** [review]
- [x] Read changes
- [x] Compare functionality of changes with Rose 1.
- [x] Picked out any gross errors in CLI.
- [x] Created issues for out of scope problems spotted:
  - #2499 
  - #2500 
  - #2501 
  - #2502

File: metomi/rose/metadata_graph.py [hunk 16]
```diff
@@ -294,9 +294,46 @@ def main():
     """Run the metadata graphing from the command line."""
     _load_override_config()
     metomi.rose.macro.add_meta_paths()
-    opt_parser = metomi.rose.opt_parse.RoseOptionParser()
-    options = ["conf_dir", "meta_path", "output_dir", "property"]
-    opt_parser.add_my_options(*options)
+    opt_parser = metomi.rose.opt_parse.RoseOptionParser(
+        usage='rose metadata-graph [OPTIONS] [SECTION ...]',
```
Comment (datamel) [inline metomi/rose/metadata_graph.py:298]: I can't preview this cli docs as I get 
```
[~/rose]$ rose metadata-graph --help
This command has been removed pending re-implementation
```
I'm assuming this is ready in place for reimplementation.

File: metomi/rose/upgrade.py [hunk 17]
```diff
@@ -778,9 +778,35 @@ def get_meta_upgrade_module(meta_path):
...
+If an application contains optional configurations, loop through
+each one, combine with the main, upgrade it, and re-create it as
+a diff vs the upgraded main configuration.
+        ''',
+        epilog='''
+ARGUMENTS
+    VERSION
+        A version to change to. If no version is specified, show available
+        versions. If `--non-interactive` is used, use the latest version
+        available. If `--non-interactive` and `--downgrade` are used, use
```
Comment (datamel) [inline metomi/rose/upgrade.py:802]: Do we need to say what happens if you just use the `--downgrade` without the `--non-interactive`?

File: metomi/rosie/suite_id.py [hunk 18]
```diff
@@ -579,7 +579,39 @@ def to_web(self):
 
 def main():
     """Implement the "rose suite-id" command."""
```
Comment (datamel) [inline metomi/rosie/suite_id.py:581]: No command rose suite-id. Does this need to be 
```suggestion
    """Implement the "rosie id" command."""
```

File: metomi/rosie/ws.py [hunk 19]
```diff
@@ -471,7 +471,19 @@ def parse_cli(*args, **kwargs):
         None:
             bare command, requesting to print server status
     """
-    opt_parser = RoseOptionParser()
+    opt_parser = RoseOptionParser(
+        description='''
+Start/stop ad-hoc Rosie suite discovery web service server.
```
Comment (datamel) [inline metomi/rosie/ws.py:476]: Again have't been able to preview this due to disabling but assuming this is in place for reimplementation.

**Review by datamel — COMMENTED** [review]
All viewed. Minor comments. I'm happy with the changes. I've tested on cli to check commands. 
I came across one error which may need dealing with? I'll approve once this has been dealt with (if you deem it necessary to deal with, might be appropriate for a follow-up pr if you want this one in quickly).
Thanks @oliver-sanders.

File: metomi/rose/rose.py [hunk 20]
```diff
@@ -0,0 +1,323 @@
...
+    }
+}
+
+
+BASH_SUB_CMDS = {
+    # bash sub commands - hardcoded
+    # NOTE: script must exist in the bin/ directory as `{ns}-{sub_cmd}`.
+    # (ns, sub_cmd)
+    ('rosa', 'db-create'),
+    ('rosa', 'rpmbuild'),
```
[the remark this answers is not in the document]
Comment (datamel) [reply metomi/rose/rose.py:118]: Ah woops, I should have looked more closely at @wxtim issues that he created.

**Review by datamel — APPROVED** [review]
Issue opened for error found. Approving and merging.

File: metomi/rose/upgrade.py (hunk 17 above)
Comment (datamel) [reply metomi/rose/upgrade.py:802]: Feel free to do this on follow-up if you feel necessary @oliver-sanders. I thought it minor so merged anyway.

File: metomi/rosie/suite_id.py (hunk 18 above)
Comment (datamel) [reply metomi/rosie/suite_id.py:581]: Feel free to do this on follow-up if you feel necessary @oliver-sanders. I thought it minor (docstring) so merged anyway.

**Pull request merged**